### PR TITLE
Persistent transcripts + compaction (issue #35)

### LINKED ISSUE #35: There should just be one session per user per clarity protocol
If a user comes back to the same clarity protocol, it should always be continuing the prior conversation. This should use a combination of a transcript and chat log compaction, just like Claude Code etc does.

Advantages:
- Fast load when you come to the session -- it isn't trying to restart Clarity from scratch by default.
- Continuity of conversation

### PR DESCRIPTION
**Pull request by yonatanzunger**
Closes #35.

## Summary

The current code treats a "user session" weirdly: whenever you restart the service it starts a new transcript file, state may or may not be preserved, and when you start up it often goes to the beginning of the clarity flow, rather than resuming cleanly as users expect. This PR fixes that.

The core idea is to rethink the concept of a session and turn it into a "conversation thread" (different name to avoid conflict with other "session" concepts in the app like WebSockets sessions). Each clarity protocol has a single associated conversation thread, whose transcript is stored in transcripts/. This transcript is broken into "chapters," with only the current chapter being of interest; a new chapter starts when either the user explicitly requests that we go to a clean sheet (preserving only the protocol) and start the conversation afresh, or when we need to do a compaction to keep the transcript from becoming infinitely long.

Transcripts are owned and managed by the ChatBackend, which does three basic things with them:
- Use them to resume a session: When the ChatBackend is started up, it tries to resume from an LLM session ID (if present), and failing that, from the transcript of the current chapter. This provides that seamless continuity that's the point of this change.
- Actually generate them: the ChatBackend writes to them.
- Compact them: Since we need to be able to start from a transcript, we need to keep the size finite. For backends like ClaudeSDK or GHCP, which maintain their own state and do their own compaction, we piggyback off that, receiving compaction events from the LLM and creating new transcript chapters to match so that we're always in sync. For backends that use a "raw" LLM (ie ones that use ClientChatBackend), we do the compaction ourselves.

Legacy transcript files are silently migrated to the new format.

From a UX perspective, there are two changes: Startups are now fast and smooth, and the weird "new session" button is replaced with "new chapter" that actually does something useful.

## Test plan

- [x] Unit suite: 1284 passed, 1 skipped — covers transcript mechanics, event/migration edge cases, compaction triggers (both backend-signaled and threshold-driven), session-manager bridge wiring, and resume-mode forks.
- [ ] Open a fresh project, chat, close the app, reopen — conversation continues.
- [ ] Same flow with a non-Claude provider (Copilot or an API client) to exercise the transcript-rebuild resume path.
- [ ] Drive a long conversation past the 85% threshold on a `ClientChatBackend` provider and verify the "Summarizing…" status, the `compaction_complete` banner, and the new chapter on disk.
- [ ] On a Claude SDK project, trigger provider-side compaction (long chat) and verify the SDK's summary is recorded in the transcript verbatim with no extra LLM call from us.
- [ ] Click "New chapter" and confirm the rollover clears the SDK session, archives the old chapter, and starts the next turn from a clean slate.
- [ ] Open a project that has pre-#35 timestamped transcript files and verify the legacy migration folds them into chapter 1.

### CODE REVIEW

**Comment by EHO-makai** [comment]
I'll checkout this branch and run a test via AOAI as another "back end" test.

**Comment by EHO-makai** [comment]
one minor comment:

ToolUse/ToolResult split during compaction. The 70/30 split in compact() filters on "message-shaped events" but could theoretically split between a ToolUse and its ToolResult. When ClientChatBackend rebuilds conversation_history from anthropic_messages(), an orphaned tool_result without a matching tool_use would cause an API error. The risk is low (paired events are usually adjacent and the 30% tail is generous), but the code doesn't explicitly guard against it. Worth adding a targeted test case and possibly a guard that walks the split point backward to include any unpaired tool use.

**Comment by EHO-makai** [comment]
Testing: I shutdown and restarted and the current chapter came back clean and exactly the same point, exactly what I was / we were wanting.

File: src/clarity_agent/transcript/transcript.py [hunk 1]
```diff
@@ -0,0 +1,966 @@
...
+class Transcript:
+    """The conversation thread for one project.
+
+    See module docstring for the conceptual model (chapters, file
+    pairs, source-of-truth invariants) and lifecycle pattern
+    (construction is cheap; file handles only exist while open;
+    use as a context manager or call :meth:`close` explicitly).
+    """
+
+    def __init__(self, project_dir: Path | str) -> None:
```
Comment (EHO-makai) [inline src/clarity_agent/transcript/transcript.py:279]: Transcript.__init__ runs migration check on every construction. [GET /api/thread/current-chapter] constructs a fresh [Transcript] per request → one call each time. Fine for typical use, but consider caching or a one-time check for projects with many chapters.

File: src/clarity_agent/transcript/transcript.py [hunk 2]
```diff
@@ -0,0 +1,966 @@
...
+        compaction pipeline).  Produces the structured
+        ``messages.create(messages=...)`` shape with ``tool_use`` /
+        ``tool_result`` blocks preserved and metadata events skipped.
+
+        Returns an empty list when the current chapter has no
+        message-producing events.
+        """
+        return build_anthropic_messages(self.current_events())
+
+    def estimated_token_count(self) -> int:
```
Comment (EHO-makai) [inline src/clarity_agent/transcript/transcript.py:742]: estimated_token_count uses chars / 4 — underestimates for CJK text, could delay compaction for non-English projects.

**Comment by EHO-makai** [comment]
nit: System prompt duplication across three backends (ClientChatBackend, SdkChatBackend, CopilotChatBackend) — extract to a shared helper.

**Review by EHO-makai — APPROVED** [review]

File: src/clarity_agent/transcript/transcript.py (hunk 1 above)
Comment (yonatanzunger) [reply src/clarity_agent/transcript/transcript.py:279]: This migration code should go away in a week or two anyway, it's not like there are a bunch of existing projects sitting around. :)
